File: Dynamic Programming/triangle.py
from typing import List
# 136ms 7.01%
# 29.5MB 5.07%
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if len(triangle[0]) == 0:
            return 0
        for i in range(1, len(triangle)):
            for j in range(len(triangle[i])):
                boundary = len(triangle[i]) -1
                if j == 0:
                    triangle[i][j] += triangle[i-1][j]
                elif j == boundary:
                    triangle[i][j] += triangle[i-1][j-1]
                else:
                    triangle[i][j] += min(triangle[i-1][j], triangle[i-1][j-1])
        return min(triangle[-1])

a = Solution()
l = [
     []
]
print(a.minimumTotal(l))

File: Dynamic Programming/triangle.cpp
#include <bits/stdc++.h>
using namespace std;

// class Solution {
// public:
//     int minimumTotal(vector<vector<int>>& triangle) {
//         //two-dimensional dp array
//         vector<vector<int>> dp;
//         // base case
//         dp.push_back({triangle[0][0]});

//         for (int i=1; i<triangle.size(); i++){
//             dp.push_back(vector<int>(i+1, 0));
//             for (int j=0; j<i+1; j++){
//                 if (j==0) dp[i][j] = dp[i-1][j] + triangle[i][j];
//                 else if (j==i) dp[i][j] = dp[i-1][j-1] + triangle[i][j];
//                 else dp[i][j] = min(dp[i-1][j], dp[i-1][j-1]) + triangle[i][j];
//             }
//         }
//         return *min_element(dp.back().begin(), dp.back().end());
//     }
// };

class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        //one-dimensional dp array
        vector<int> dp;
        dp.reserve(triangle.size());
        // base case
        dp.push_back(triangle[0][0]);

        for (int i=1; i<triangle.size(); i++){
            dp.push_back(0);
            int temp = dp[0];
            dp[0] = dp[0] + triangle[i][0];
            for (int j=1; j<i+1; j++){
                int temp2 = dp[j];
                if (j==i) dp[j] = temp + triangle[i][j];
                else dp[j] = min(dp[j], temp) + triangle[i][j];
                temp = temp2;
            }
        }
        return *min_element(dp.begin(), dp.end());
    }
};

int main(){
    Solution sol;
    vector<vector<int>> triangle = {{-10}, {1, 2}};
    cout << sol.minimumTotal(triangle) << endl;
    return 0;
}